File: cmd/specs/bootstrap.go
package main

import (
	"flag"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"

	"github.com/jdehaan/specs-cli/internal/tools"
)

// cmdBootstrap scaffolds a new host with .specs.yaml pointing at the
// framework content. The default mode is `managed`: content is fetched
// once into the user's cache dir and shared across host projects.
//
//	--at .                 makes the cwd itself the specs root
//	--layout folder        creates specs/ as a plain folder (default)
//	--layout submodule     register specs/ as a submodule (advanced; manual)
//
//	--tools-mode managed   (default) fetch into the user cache, hide it
//	--tools-mode submodule add .specs-tools as a submodule of the host
//	--tools-mode folder    clone .specs-tools next to specs root
//	--tools-mode vendor    snapshot .specs-tools (no .git)
func cmdBootstrap(args []string) error {
	fs := flag.NewFlagSet("bootstrap", flag.ContinueOnError)
	at := fs.String("at", "", "path to the specs root (created if missing); use '.' for repo root")
	layout := fs.String("layout", "folder", "how specs/ is materialised: submodule|folder")
	toolsMode := fs.String("tools-mode", "managed", "how .specs-tools is materialised: managed|submodule|folder|vendor")
	toolsURL := fs.String("tools-url", "https://github.com/jdehaan/specs-tools.git", "git URL of specs-tools content repo")
	toolsRef := fs.String("tools-ref", "main", "tag/branch/commit for content")
	withModel := fs.Bool("with-model", false, "create empty model/ and change-requests/ skeletons")
	withVSCode := fs.Bool("with-vscode", false, "write .vscode/tasks.json")
	dryRun := fs.Bool("dry-run", false, "print actions without performing them")
	fs.Usage = func() {
		fmt.Fprintln(os.Stderr, "Usage: specs bootstrap [--at <path>] [--layout submodule|folder] [--tools-mode managed|submodule|folder|vendor] [--tools-url URL] [--tools-ref REF] [--with-model] [--with-vscode] [--dry-run]")
		fs.PrintDefaults()
	}
	if err := fs.Parse(args); err != nil {
		return err
	}

	cwd, err := os.Getwd()
	if err != nil {
		return err
	}

	// Resolve specs root path.
	var specsRoot string
	switch *at {
	case "":
		specsRoot = filepath.Join(cwd, "specs")
	case ".":
		specsRoot = cwd
	default:
		if filepath.IsAbs(*at) {
			specsRoot = *at
		} else {
			specsRoot = filepath.Join(cwd, *at)
		}
	}

	switch *layout {
	case "folder":
		if err := runOrLog(*dryRun, "mkdir -p", func() error { return os.MkdirAll(specsRoot, 0o755) }); err != nil {
			return err
		}
	case "submodule":
		// Add specs root as a submodule of cwd. URL not collected here (operator
		// adds the upstream specs repo separately or via --tools-url-style fork).
		return exitWith(2, "--layout submodule requires the host to register the submodule explicitly first; see docs")
	default:
		return exitWith(2, "unknown --layout %q", *layout)
	}

	// Materialise .specs-tools content.
	toolsDir := filepath.Join(specsRoot, ".specs-tools")
	switch *toolsMode {
	case "managed":
		if *dryRun {
			fmt.Printf("would: fetch %s@%s into managed cache\n", *toolsURL, *toolsRef)
		} else {
			path, err := tools.Ensure(*toolsURL, *toolsRef)
			if err != nil {
				return exitWith(1, "fetch managed tools: %v", err)
			}
			fmt.Printf("managed tools cached at %s\n", path)
		}
	case "submodule":
		// Submodule must be added in the host repo's git root, not below specsRoot
		// when specsRoot is itself just a folder. Auto-detect the git root from cwd.
		hostGitRoot := findGitRoot(specsRoot)
		if hostGitRoot == "" {
			// initialise a git repo in cwd if none exists, so we can register a submodule
			if err := runOrLog(*dryRun, "git init "+cwd, func() error {
				return runGit(cwd, "init")
			}); err != nil {
				return err
			}
			hostGitRoot = cwd
		}
		rel, _ := filepath.Rel(hostGitRoot, toolsDir)
		gitArgs := []string{"submodule", "add"}
		if *toolsRef != "" {
			gitArgs = append(gitArgs, "-b", *toolsRef)
		}
		gitArgs = append(gitArgs, *toolsURL, rel)
		if err := runOrLog(*dryRun, fmt.Sprintf("git -C %s %v", hostGitRoot, gitArgs), func() error {
			return runGit(hostGitRoot, gitArgs...)
		}); err != nil {
			return err
		}
	case "folder":
		gitArgs := []string{"clone"}
		if *toolsRef != "" {
			gitArgs = append(gitArgs, "--branch", *toolsRef)
		}
		gitArgs = append(gitArgs, *toolsURL, toolsDir)
		if err := runOrLog(*dryRun, fmt.Sprintf("git %v", gitArgs), func() error {
			return runGit("", gitArgs...)
		}); err != nil {
			return err
		}
	case "vendor":
		// Vendor: same as folder, but strip the .git directory afterwards.
		gitArgs := []string{"clone", "--depth", "1"}
		if *toolsRef != "" {
			gitArgs = append(gitArgs, "--branch", *toolsRef)
		}
		gitArgs = append(gitArgs, *toolsURL, toolsDir)
		if err := runOrLog(*dryRun, fmt.Sprintf("git %v && rm -rf %s/.git", gitArgs, toolsDir), func() error {
			if err := runGit("", gitArgs...); err != nil {
				return err
			}
			return os.RemoveAll(filepath.Join(toolsDir, ".git"))
		}); err != nil {
			return err
		}
	default:
		return exitWith(2, "unknown --tools-mode %q", *toolsMode)
	}

	if *withModel {
		for _, sub := range []string{"model", "change-requests"} {
			p := filepath.Join(specsRoot, sub)
			if err := runOrLog(*dryRun, "mkdir -p "+p, func() error { return os.MkdirAll(p, 0o755) }); err != nil {
				return err
			}
		}
	}

	// Run init logic to write .specs.yaml in the new specs root.
	initArgs := []string{"--at", specsRoot, "--force"}
	if *toolsMode == "managed" {
		initArgs = append(initArgs, "--tools-url", *toolsURL, "--tools-ref", *toolsRef)
	}
	if *withVSCode {
		initArgs = append(initArgs, "--with-vscode")
	}
	if *dryRun {
		fmt.Printf("would: specs init %v\n", initArgs)
		return nil
	}
	return cmdInit(initArgs)
}

func runOrLog(dryRun bool, label string, fn func() error) error {
	if dryRun {
		fmt.Println("would:", label)
		return nil
	}
	fmt.Println("$", label)
	return fn()
}

func runGit(dir string, args ...string) error {
	cmd := exec.Command("git", args...)
	if dir != "" {
		cmd.Dir = dir
	}
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
	return cmd.Run()
}

func findGitRoot(start string) string {
	d := start
	for {
		if _, err := os.Stat(filepath.Join(d, ".git")); err == nil {
			return d
		}
		parent := filepath.Dir(d)
		if parent == d {
			return ""
		}
		d = parent
	}
}

File: extension/src/bootstrap.ts
// Phase E2 — bootstrap wizard. Maps multi-step QuickPick answers to a
// 'specs bootstrap ...' invocation.
import * as vscode from "vscode";
import { runInTerminal, runAndCapture, getOutput } from "./cli";

interface BootstrapAnswers {
    layout: "folder" | "submodule";
    specsUrl?: string;
    specsRef?: string;
    toolsMode: "managed" | "submodule" | "folder" | "vendor";
    toolsUrl: string;
    toolsRef: string;
    withModel: boolean;
    withVscode: boolean;
}

const DEFAULT_TOOLS_URL = "https://github.com/jdehaan/specs-tools.git";

export async function runBootstrapWizard(
    context: vscode.ExtensionContext
): Promise<void> {
    const folder = pickFolder();
    if (!folder) {
        return;
    }

    const layout = await pickLayout();
    if (!layout) {
        return;
    }

    let specsUrl: string | undefined;
    let specsRef: string | undefined;
    if (layout === "submodule") {
        specsUrl = await vscode.window.showInputBox({
            prompt: "Git URL of the host's specs repo (--specs-url)",
            placeHolder: "https://github.com/<org>/<host-specs>.git",
            ignoreFocusOut: true,
            validateInput: (v) =>
                v.trim().length === 0 ? "URL is required for submodule layout" : null,
        });
        if (!specsUrl) {
            return;
        }
        specsRef = await vscode.window.showInputBox({
            prompt: "Branch/tag for the specs submodule (optional)",
            placeHolder: "main",
            ignoreFocusOut: true,
        });
    }

    const toolsMode = await pickToolsMode();
    if (!toolsMode) {
        return;
    }

    const toolsUrl = await vscode.window.showInputBox({
        prompt: "Tools content git URL (--tools-url)",
        value: DEFAULT_TOOLS_URL,
        ignoreFocusOut: true,
    });
    if (toolsUrl === undefined) {
        return;
    }
    const toolsRef = await vscode.window.showInputBox({
        prompt: "Tools content ref (--tools-ref)",
        value: "main",
        ignoreFocusOut: true,
    });
    if (toolsRef === undefined) {
        return;
    }

    const extras = await vscode.window.showQuickPick(
        [
            { label: "Create model/ and change-requests/ skeletons", picked: true },
            { label: "Write .vscode/tasks.json", picked: false },
        ],
        {
            canPickMany: true,
            placeHolder: "Optional scaffolding",
            ignoreFocusOut: true,
        }
    );
    if (!extras) {
        return;
    }
    const withModel = extras.some((e) => e.label.startsWith("Create model"));
    const withVscode = extras.some((e) => e.label.startsWith("Write .vscode"));

    const answers: BootstrapAnswers = {
        layout,
        specsUrl,
        specsRef,
        toolsMode,
        toolsUrl,
        toolsRef,
        withModel,
        withVscode,
    };

    const args = buildArgs(answers);

    // Show dry-run preview first.
    const out = getOutput();
    out.show(true);
    out.appendLine("Specs bootstrap (dry-run preview)");
    const preview = await runAndCapture(context, [...args, "--dry-run"], folder.uri.fsPath);
    out.appendLine(preview.stdout);
    if (preview.stderr) {
        out.appendLine(preview.stderr);
    }
    if (preview.exitCode !== 0) {
        vscode.window.showErrorMessage(
            `bootstrap dry-run failed (exit ${preview.exitCode}). See Specs output.`
        );
        return;
    }

    const choice = await vscode.window.showInformationMessage(
        `Run 'specs ${args.join(" ")}' in ${folder.uri.fsPath}?`,
        { modal: true },
        "Run",
        "Cancel"
    );
    if (choice !== "Run") {
        return;
    }
    runInTerminal(context, args, folder.uri.fsPath, "Specs: bootstrap");
}

function buildArgs(a: BootstrapAnswers): string[] {
    const args = ["bootstrap", "--layout", a.layout];
    if (a.layout === "submodule" && a.specsUrl) {
        args.push("--specs-url", a.specsUrl);
        if (a.specsRef) {
            args.push("--specs-ref", a.specsRef);
        }
    }
    args.push("--tools-mode", a.toolsMode);
    if (a.toolsUrl) {
        args.push("--tools-url", a.toolsUrl);
    }
    if (a.toolsRef) {
        args.push("--tools-ref", a.toolsRef);
    }
    if (a.withModel) {
        args.push("--with-model");
    }
    if (a.withVscode) {
        args.push("--with-vscode");
    }
    return args;
}

function pickFolder(): vscode.WorkspaceFolder | undefined {
    const folders = vscode.workspace.workspaceFolders ?? [];
    if (folders.length === 0) {
        vscode.window.showWarningMessage(
            "Specs: open a folder first; bootstrap operates on the active workspace."
        );
        return undefined;
    }
    if (folders.length === 1) {
        return folders[0];
    }
    // Multi-root: would require a separate prompt; defer for now.
    vscode.window.showWarningMessage(
        "Specs: multi-root workspaces are not yet supported by the bootstrap wizard."
    );
    return undefined;
}

async function pickLayout(): Promise<BootstrapAnswers["layout"] | undefined> {
    const items: vscode.QuickPickItem[] = [
        {
            label: "Folder",
            description: "specs/ is a plain folder in the host repo (recommended for new repos)",
        },
        {
            label: "Submodule",
            description: "specs/ is a git submodule of an existing specs repo",
        },
    ];
    const pick = await vscode.window.showQuickPick(items, {
        placeHolder: "How should specs/ be materialised?",
        ignoreFocusOut: true,
    });
    if (!pick) {
        return undefined;
    }
    return pick.label === "Submodule" ? "submodule" : "folder";
}

async function pickToolsMode(): Promise<BootstrapAnswers["toolsMode"] | undefined> {
    const items: vscode.QuickPickItem[] = [
        {
            label: "managed",
            description: "fetch into the user cache, share across projects (recommended)",
        },
        { label: "submodule", description: "add .specs-tools as a submodule of the host repo" },
        { label: "folder", description: "clone .specs-tools next to the specs root" },
        { label: "vendor", description: "snapshot .specs-tools without git history" },
    ];
    const pick = await vscode.window.showQuickPick(items, {
        placeHolder: "How should .specs-tools be materialised?",
        ignoreFocusOut: true,
    });
    if (!pick) {
        return undefined;
    }
    return pick.label as BootstrapAnswers["toolsMode"];
}
